connector/saml/saml.go

```diff
@@ -0,0 +1,171 @@
+	}
+	return idpc, nil
+}
+
+var (
+	_ connector.CallbackConnector = (*samlConnector)(nil)
+)
+
+type samlConnector struct {
+	spSettings              saml.ServiceProviderSettings
+	spNameIDPolicyFormat    string
+	spRequestedAuthnContext string
+	idAttr                  string
+	userAttr                string
+	emailAttr               string
+	groupsAttr              string
+	trustedEmailProvider    bool
+}
+
+func (c *samlConnector) Close() error {
+	return nil
+}
+
+func (c *samlConnector) LoginURL(s connector.Scopes, callbackURL, state string) (string, error) {
+	// Construct an AuthnRequest
+	authnRequest := c.spSettings.GetAuthnRequest()
+
+	if len(c.spNameIDPolicyFormat) != 0 {
+		authnRequest.NameIDPolicy.Format = c.spNameIDPolicyFormat
+	}
+
+	if len(c.spRequestedAuthnContext) == 0 {
+		authnRequest.RequestedAuthnContext = nil
+	}
+
+	b64XML, err := authnRequest.CompressedEncodedString()
+	if err != nil {
+		return "", err
+	}
+
+	// Get a URL formed with the SAMLRequest parameter
+	url, err := saml.GetAuthnRequestURL(c.spSettings.IDPSSOURL, b64XML, state)
+	if err != nil {
+		return "", err
+	}
+
+	if c.spSettings.AssertionConsumerServiceURL != callbackURL {
+		return "", fmt.Errorf("expected callback URL did not match the URL in the config")
+	}
+	return url, err
+}
+
+type samlError struct {
+	error            string
+	errorDescription string
+}
+
+func (e *samlError) Error() string {
+	if e.errorDescription == "" {
+		return e.error
+	}
+	return e.error + ": " + e.errorDescription
+}
+
+func (c *samlConnector) HandleCallback(s connector.Scopes, r *http.Request) (identity connector.Identity, err error) {
+	q := r.URL.Query()
+	if errType := q.Get("error"); errType != "" {
+		return identity, &samlError{errType, q.Get("error_description")}
+	}
+
+	encodedXML := r.FormValue("SAMLResponse")
+	if encodedXML == "" {
+		return identity, fmt.Errorf("saml: SAMLResponse form value missing")
+	}
+
+	response, err := saml.ParseEncodedResponse(encodedXML)
+	if err != nil {
+		return identity, fmt.Errorf("saml: SAMLResponse parse: %v", err.Error())
+	}
+
+	err = response.Validate(&c.spSettings)
+	if err != nil {
+		return identity, fmt.Errorf("saml: SAMLResponse validation: %v", err.Error())
+	}
+
+	userID := response.GetAttribute(c.idAttr)
+	if userID == "" {
+		return identity, fmt.Errorf("saml: SAML attribute identifier %s missing", c.idAttr)
+	}
+
+	username := response.GetAttribute(c.userAttr)
+	if username == "" {
+		return identity, fmt.Errorf("saml: SAML attribute identifier %s missing", c.userAttr)
+	}
+
+	email := response.GetAttribute(c.emailAttr)
+	if email == "" {
+		return identity, fmt.Errorf("saml: SAML attribute identifier %s missing", c.emailAttr)
+	}
+
+	identity = connector.Identity{
+		UserID:        userID,
+		Username:      username,
+		Email:         email,
+		EmailVerified: c.trustedEmailProvider,
+	}
+
+	if s.Groups {
+		groups := response.GetAttribute(c.groupsAttr)
+		identity.Groups = strings.Split(groups, ",")
```

Comment:
I can see this behavior with our SSO implementation (Oracle Access Manager), but can't confirm that this behavior is consistent among others.